File: code/0160 Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    int lA=0,lB=0;  //Length of the list
    for(struct ListNode* pA=headA; pA!=NULL; pA=pA->next) lA++;
    for(struct ListNode* pB=headB; pB!=NULL; pB=pB->next) lB++;
    
    /* move the longer one forward to match two lists at the start-line */
    for(int i = 0; i<(lA-lB); i++) headA=headA->next;
    for(int i = 0; i<(lB-lA); i++) headB=headB->next;

    while(headA!=headB)
    {
        headA=headA->next;
        headB=headB->next;
    }
    return headA;
}

File: Python/0160 Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA is None or headB is None: return None
        ca,cb = headA, headB
        for _ in range(3):
            while(ca and cb):
                if ca == cb: return ca
                ca = ca.next
                cb = cb.next
            if ca is None: ca = headB
            if cb is None: cb = headA        
        return None
